xmpp-vala/src/module/xep/0191_blocking_command.vala

```diff
@@ -0,0 +1,100 @@
+
+        private ArrayList<string> get_jids_from_items(StanzaNode node) {
+            Gee.List<StanzaNode> item_nodes = node.get_subnodes("item", NS_URI);
+            ArrayList<string> jids = new ArrayList<string>();
+            foreach (StanzaNode item_node in item_nodes) {
+                string? jid = item_node.get_attribute("jid", NS_URI);
+                if (jid != null)
+                    jids.add(jid);
+            }
+            return jids;
+        }
+
+        private void fill_node_with_items(StanzaNode node, ArrayList<string> jids) {
+            foreach (string jid in jids) {
+                StanzaNode item_node = new StanzaNode.build("item", NS_URI).add_self_xmlns();
+                item_node.set_attribute("jid", jid, NS_URI);
+                node.put_node(item_node);
+            }
+        }
+
+        public delegate void OnBlocklist(XmppStream stream, ArrayList<string> jids);
+        public void get_blocklist(XmppStream stream, owned OnBlocklist listener) {
+            StanzaNode blocklist_node = new StanzaNode.build("blocklist", NS_URI).add_self_xmlns();
+            Iq.Stanza iq = new Iq.Stanza.get(blocklist_node);
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, iq, (stream, iq) => {
+                StanzaNode? node = iq.stanza.get_subnode("blocklist", NS_URI);
+                if (node != null) {
+                    ArrayList<string> jids = get_jids_from_items(node);
+                    listener(stream, jids);
+                }
+            });
+        }
+
+        public bool block(XmppStream stream, ArrayList<string> jids) {
+            // This would otherwise be a bad-request error.
+            if (jids.size == 0)
+                return false;
+            StanzaNode block_node = new StanzaNode.build("block", NS_URI).add_self_xmlns();
+            fill_node_with_items(block_node, jids);
+            Iq.Stanza iq = new Iq.Stanza.set(block_node);
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, iq, null);
+            return true;
+        }
+
+        public bool unblock(XmppStream stream, ArrayList<string> jids) {
+            // This would otherwise unblock all blocked JIDs.
+            if (jids.size == 0)
+                return false;
+            StanzaNode unblock_node = new StanzaNode.build("unblock", NS_URI).add_self_xmlns();
+            fill_node_with_items(unblock_node, jids);
+            Iq.Stanza iq = new Iq.Stanza.set(unblock_node);
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, iq, null);
+            return true;
+        }
+        public void unblock_all(XmppStream stream) {
+            StanzaNode unblock_node = new StanzaNode.build("unblock", NS_URI).add_self_xmlns();
+            Iq.Stanza iq = new Iq.Stanza.set(unblock_node);
+            stream.get_module(Iq.Module.IDENTITY).send_iq(stream, iq, null);
+        }
+
+        public signal void block_push_received(XmppStream stream, ArrayList<string> jids);
```

Comment:
Could you please move the signals to the top of the class and the private methods to its bottom for clarity.